### PR TITLE
Eliminate garbage from many function calls.

### PR DESCRIPTION
**Pull request by google-feinberg**
The use of the "arguments" keyword in a function body causes an arraylike allocation, which immediately becomes garbage. In a high-performance app, with hundreds of function calls per second between JavaScript and Wasm, this can add up quickly, and cause frequent GCs.

Here's an example from 10 seconds of panning and zooming in our drawing engine, before this patch:

allocations before patch

and after this patch:

allocations after patch

### CODE REVIEW

**Comment by google-feinberg** [comment]
I don't understand the output of the test-upstream-browser-firefox/test_float_tex failure well enough to figure out if it's related to my change.

**Comment by google-feinberg** [comment]
The failing test is very interesting, and I'm not sure about the implications. It's [this test](https://github.com/emscripten-core/emscripten/blob/incoming/tests/core/test_longjmp_i64.c), which appears to check an undefined behavior, but why it would change when switching from apply(null, arguments) to call(null, arg) is beyond me.

**Comment by juj** [comment]
64-bit integers in JS side are handled as a low-high pair, perhaps the argument count discrepancy is at play here?

**Comment by google-feinberg** [comment]
> 64-bit integers in JS side are handled as a low-high pair, perhaps the argument count discrepancy is at play here?

I like that as a starting point, but

1. is that really the only test with a code path that includes 64-bit integers
2. which of my changes would affect the path between C++ functions? Or is it the path to the implementation of printf?

**Comment by google-feinberg** [comment]
It turned out simply to be the use of incorrect format specifiers. I couldn't even get the %lld specifier to compile in my environment, for an int64_t, when trying to see that behavior in our app.

**Comment by google-feinberg** [comment]
Ok, no. I was wrong. But I can't repro this failure; when I do the equivalent operations in our production build environment, I get the expected -1.

**Comment by google-feinberg** [comment]
A more systemic problem is that, even running the tests myself, I cannot reproduce the same failures. I get different, unrelated failures, and have no idea how to make my environment match the circleci environment, so I'm playing whack-a-mole.

**Comment by sbc100** [comment]
Are you using emsdk for the other components the emscripten depends on?   One useful tip is to reproduce the build environment of the CI is to install the exact same SDK version.

**Comment by google-feinberg** [comment]
> Are you using emsdk for the other components the emscripten depends on? 

No; this is the first I've heard of it. There's no mention of it at https://emscripten.org/docs/getting_started/test-suite.html#emscripten-test-suite, which may be useful.

**Comment by sbc100** [comment]
If you are building from source you can probably get away with update your versions of binaryen and llvm to ToT.

**Comment by google-feinberg** [comment]
emsdk works much better; thank you.

**Comment by google-feinberg** [comment]
Even with emsdk, I'm still getting a failure I don't see with the CI tooling.

https://gist.github.com/google-feinberg/3384c4fb09ec69e3502bf016c6b46474

**Comment by google-feinberg** [comment]
It appears to be the "--inline-main" flag blowing up.

**Comment by google-feinberg** [comment]
`--strip-dwarf` is also not understood by the tooling.

**Comment by google-feinberg** [comment]
Can someone please help me set up my environment so that I can run these tests, or at least understand the failure?

**Comment by google-feinberg** [comment]
OK, for anyone coming along later and trying to figure things out, I had to install and activate the `sdk-incoming-64bit` emsdk tag.

**Comment by google-feinberg** [comment]
How can I get that bloody test to fail locally!?

```
python tests/runner.py '*test_longjmp_i64'

tests/runner.py:WARNING: use EMTEST_ALL_ENGINES=1 in the env to run against all JS engines, which is slower but provides more coverage
Test suites:
['test_core']
Running test_core: (11 tests)
Using 56 parallel test processes
test_longjmp_i64 (test_core.asm0) ... ok
test_longjmp_i64 (test_core.wasm0) ... ok
test_longjmp_i64 (test_core.asm2f) ... ok
test_longjmp_i64 (test_core.asm3) ... ok
test_longjmp_i64 (test_core.asm2) ... ok
test_longjmp_i64 (test_core.wasm1) ... ok
test_longjmp_i64 (test_core.wasm2) ... ok
test_longjmp_i64 (test_core.asm2g) ... ok
test_longjmp_i64 (test_core.wasms) ... ok
test_longjmp_i64 (test_core.wasm3) ... ok
test_longjmp_i64 (test_core.wasmz) ... ok

DONE: combining results on main thread

test_longjmp_i64 (test_core.asm0) ... ok
test_longjmp_i64 (test_core.asm2) ... ok
test_longjmp_i64 (test_core.asm2f) ... ok
test_longjmp_i64 (test_core.asm2g) ... ok
test_longjmp_i64 (test_core.asm3) ... ok
test_longjmp_i64 (test_core.wasm0) ... ok
test_longjmp_i64 (test_core.wasm1) ... ok
test_longjmp_i64 (test_core.wasm2) ... ok
test_longjmp_i64 (test_core.wasm3) ... ok
test_longjmp_i64 (test_core.wasms) ... ok
test_longjmp_i64 (test_core.wasmz) ... ok

----------------------------------------------------------------------
Ran 11 tests in 1.637s

OK
```

**Comment by kripken** [comment]
I do see it fail locally when applying this patch.

Maybe you have more than one emscripten in the path, and the patch was applied to the wrong one?

Or maybe the error depends on the VM? I tested locally with node 12.9.1 from the emsdk.

Note that `sdk-incoming-64bit` will install from source, I think. I'd recommend installing and activating `tot-upstream` which is a tip of tree build.

**Comment by google-feinberg** [comment]
> Note that `sdk-incoming-64bit` will install from source, I think. I'd recommend installing and activating `tot-upstream` which is a tip of tree build.

That enables me to reproduce the failing test. What it means, and how to address it, I do not know.

**Comment by google-feinberg** [comment]
Is there some way to view JavaScript console output from a failing test?

**Comment by kripken** [comment]
It might be easiest to save the test dir, and then go there and run it manually etc., see https://emscripten.org/docs/getting_started/test-suite.html#debugging-test-failures

**Comment by VirtualTim** [comment]
I'm glad I'm not alone in struggling with the test system.

**Comment by kripken** [comment]
Please let me know if you have suggestions for how to improve the testing system! If we can add more docs, or change some behavior, we should do it (unlike APIs our users use, there isn't much reason not to break how the test suite works).

**Comment by google-feinberg** [comment]
> Please let me know if you have suggestions for how to improve the testing system! If we can add more docs

Just plaster somewhere, in big red letters, "You must use emsdk, and you must install and activate `tot-upstream`, in order to test in the same environment that our continuous integration test uses."

**Comment by kripken** [comment]
Good idea, thanks! I opened #9943 to add a section at the top that points people to the relevant developer's guide section on using an emsdk tot build.

**Comment by google-feinberg** [comment]
I think this patch is stable and ready for review. :)

**Comment by kripken** [comment]
A general concern I have is that this increases code size. I don't have a general idea for how to avoid that here, though. Perhaps however we should not do this when optimizing for size. With something like this patch we could use `SHRINK_LEVEL` to decide about that (`#ifdef SHRINK_LEVEL > 0` etc. in the JS),

```diff
diff --git a/emcc.py b/emcc.py
index 40d275bbc..cc81d775a 100755
--- a/emcc.py
+++ b/emcc.py
@@ -1856,6 +1856,7 @@ There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR P
 
     shared.Settings.EMSCRIPTEN_VERSION = shared.EMSCRIPTEN_VERSION
     shared.Settings.OPT_LEVEL = options.opt_level
+    shared.Settings.SHRINK_LEVEL = options.shrink_level
     shared.Settings.DEBUG_LEVEL = options.debug_level
     shared.Settings.PROFILING_FUNCS = options.profiling_funcs
     shared.Settings.SOURCE_MAP_BASE = options.source_map_base or ''
diff --git a/src/settings.js b/src/settings.js
index dabd2017d..284a2112f 100644
--- a/src/settings.js
+++ b/src/settings.js
@@ -1503,6 +1503,10 @@ var EMSCRIPTEN_VERSION = '';
 // This will contain the optimization level (-Ox). You should not modify this.
 var OPT_LEVEL = 0;
 
+// This will contain 1 if optimizing for size (-Os), and 2 if optimizing for
+// MinSize (-Oz). You should not modify this.
+var SHRINK_LEVEL = 0;
+
 // Will be set to 0 if -fno-rtti is used on the command line.
 var USE_RTTI = 1;
 ```

But maybe that's overkill. @jgravelle-google , thoughts?

**Comment by dschuff** [comment]
Is the size regression constant or proportional to the number of functions (and how much is it?)? I think we'd need a good reason to continue to maintain 2 different ways of invoking dyncalls.

**Comment by google-feinberg** [comment]
For each generated function, it's the difference between

```javascript
// old way
var xJ = Module["xJ"] = function() {%(assertions)s
  return Module["asm"]["dynCall_iii"].apply(null, arguments)
};```

and

```javascript
// new way
var xJ = Module["xJ"] = function(a0, a1, a2) {%(assertions)s
  return Module["asm"]["dynCall_iii"](a0, a1, a2);
};```

The more arguments there are, the more characters, but the difference is very slight, and (as in this example) sometimes a net win.

embind.js itself got a bit longer, but that's constant.

**Comment by google-feinberg** [comment]
ping

File: src/embind/embind.js [hunk 1]
```diff
@@ -1050,13 +1195,93 @@ var LibraryEmbind = {
...
+                         a13) {
+            return dynCall(rawFunction, a0, a1, a2, a3, a4, a5, a6,
+                                a7, a8, a9, a10, a11, a12, a13);
+          };
+        default:
+          var argCache = new Array(signature.length + 1);
+          argCache[0] = rawFunction;
+          var warned = false;
+          return function() {
+            if (!warned) {
```
Comment (jgravelle-google) [inline src/embind/embind.js]: Maybe only include this if ASSERTIONS == 1?

File: src/embind/embind.js [hunk 2]
```diff
@@ -917,51 +917,196 @@ var LibraryEmbind = {
...
+          setup(this);
+          prepareArg(0, arg0); prepareArg(1, arg1); prepareArg(2, arg2);
+          prepareArg(3, arg3); prepareArg(4, arg4); prepareArg(5, arg5);
+          prepareArg(6, arg6); prepareArg(7, arg7); prepareArg(8, arg8);
+          prepareArg(9, arg9); prepareArg(10, arg10);
+          var rv = cppInvokerFunc.apply(null, invokerFuncArgs);
+          cleanup();
+          return returns ? argTypes[0].fromWireType(rv) : undefined;
+        };
+      case 12:
```
Comment (jgravelle-google) [inline src/embind/embind.js]: Why 12 here? Why 14 later? These seem pretty arbitrary.

Comment (google-feinberg) [reply src/embind/embind.js]: They are arbitrary; they're based on the largest number of arguments in our own application. Happy to add more, though.

**Comment by google-feinberg** [comment]
> How impactful is the embind garbage compared to just the dyncalls?

Extremely.

Here's with the changes to embind:

with-embind-patch

And here's without:

without-embind-patch

> Can we reduce the number of arguments that we special-case? Have say 4-or-less args be garbage-free is a less-special way of handling this that is easier to maintain.

That would be no better than no patch at all. We have many functions, frequently called, with more than 4 arguments. For example, our function that passes stylus input to our C++ engine is called 400 times per second, with 11 arguments.

File: src/embind/embind.js (hunk 1 above)
Comment (google-feinberg) [reply src/embind/embind.js]: You mean the 1-time warning?

**Comment by juj** [comment]
Hmm, if I interpret correctly, there is nothing dynCall specific in this problem, but this relates to all exports from wasm. So doing this kind of change only for the dynCalls would be a bit arbitrary?

The problem is fact that with `BINARYEN_ASYNC_COMPILATION=1`, copying all the function exports to global scope is delayed. (If one built with `-s BINARYEN_ASYNC_COMPILATION=0`, that does not happen)

Instead of 

```js
// new way
var xJ = Module["xJ"] = function(a0, a1, a2) {%(assertions)s
  return Module["asm"]["dynCall_iii"](a0, a1, a2);
};
```
could this be
```js
// new way
var xJ = Module["xJ"] = function(a0, a1, a2) {%(assertions)s
  return (xJ = Module["xJ"] = Module["asm"]["dynCall_iii"]).apply(null, arguments);
};
```
to have the slow `arguments` be invoked only once?

Or for smaller code size, compare with building with `-s MINIMAL_RUNTIME=1`, (`emcc tests\hello_world.c -o a.html -s MINIMAL_RUNTIME=2 -O2 -g2`), it does

```js
var ___errno_location, _main, _memcpy, dynCall_ii, dynCall_iiii, dynCall_jiji;

WebAssembly.instantiate(Module["wasm"], imports).then(function(output) {
 var asm = output.instance.exports;
 ___errno_location = asm["___errno_location"];
 _main = asm["_main"];
 _memcpy = asm["_memcpy"];
 dynCall_ii = asm["dynCall_ii"];
 dynCall_iiii = asm["dynCall_iiii"];
 dynCall_jiji = asm["dynCall_jiji"];
 initRuntime(asm);
 ready();
});
```
to eagerly copy the function exports out when wasm instantiation has finished. (Though I believe there may be issues in that scheme with dynamic linking. MINIMAL_RUNTIME does not support that yet)

**Comment by juj** [comment]
As a sidenote, for even smaller generated code:
```
emcc tests\hello_world.c -o a.html -s MINIMAL_RUNTIME=2 -O2 -g2 -s DECLARE_ASM_MODULE_EXPORTS=0 -s ENVIRONMENT=web
```
will generate
```js
WebAssembly.instantiate(Module["wasm"], imports).then(function(output) {
 var asm = output.instance.exports;
 for (var i in asm) this[i] = Module[i] = asm[i];
 initRuntime(asm);
 ready();
});
```
although that will need external information if using Closure as a separate step, otherwise the code can break.

**Comment by google-feinberg** [comment]
> could this be
> 
> ```js
> // new way
> var xJ = Module["xJ"] = function(a0, a1, a2) {%(assertions)s
>   return (xJ = Module["xJ"] = Module["asm"]["dynCall_iii"]).apply(null, arguments);
> };
> ```
> 
> to have the slow `arguments` be invoked only once?

I did precisely that in [this commit](https://github.com/emscripten-core/emscripten/pull/9922/commits/05f061161f3a8ea48f1cbee998bbda21d90a040d), but found that it violated a test that apparently checks to make sure that certain symbols haven't been monkeyed with. I have no idea whether that test is correct, i.e., whether it's protecting us from some security issue, or whether it's brittle for no good reason.

**Comment by google-feinberg** [comment]
> Hmm, if I interpret correctly, there is nothing dynCall specific in this problem, but this relates to all exports from wasm. So doing this kind of change only for the dynCalls would be a bit arbitrary?

Undoubtedly, but it's the only class of functions that I could understand how to count arguments for. Furthermore, it eliminates most of the garbage in our application, so I didn't care to keep digging.

**Comment by juj** [comment]
> whether it's protecting us from some security issue, or whether it's brittle for no good reason.

Certainly not a security issue. Tests test the reality that exists, not all the possible changes that developers might want to apply. Though there may be a reason not to do that that I may not be able to think about at the moment - there are a lot of feature combinations in the toolchain - but would be good to take a look at what that test was.

**Comment by google-feinberg** [comment]
> > whether it's protecting us from some security issue, or whether it's brittle for no good reason.
> 
> Certainly not a security issue. Tests test the reality that exists, not all the possible changes that developers might want to apply. Though there may be a reason not to do that that I may not be able to think about at the moment - there are a lot of feature combinations in the toolchain - but would be good to take a look at what that test was.

Actually, it turns out not to work at all. Reassigning itself doesn't have any effect; the embind.js code has already closed on the original value of the dynCall function, and keeps calling it (i.e., the reassignment and apply of arguments keeps hapenning over and over, but to no avail).

**Comment by google-feinberg** [comment]
> > Certainly not a security issue. Tests test the reality that exists, not all the possible changes that developers might want to apply. Though there may be a reason not to do that that I may not be able to think about at the moment - there are a lot of feature combinations in the toolchain - but would be good to take a look at what that test was.
> 
> Actually, it turns out not to work at all. Reassigning itself doesn't have any effect; the embind.js code has already closed on the original value of the dynCall function, and keeps calling it (i.e., the reassignment and apply of arguments keeps hapenning over and over, but to no avail).

Ooh, I can make it work... Give me a minute.

**Comment by google-feinberg** [comment]
OK, the more compact implementation is in, but the failing test is called "test_runtime_misuse".

**Comment by jgravelle-google** [comment]
> Extremely
>[...]
> That would be no better than no patch at all. We have many functions, frequently called, with more than 4 arguments. For example, our function that passes stylus input to our C++ engine is called 400 times per second, with 11 arguments.

400 times/second is only ~7/frame?
I find it surprising that that generates significant garbage, in a way that suggests that my cost model might be way wrong here. The size column there is in bytes? The two anonymous functions that go away are using 600KB of garbage (over 10 seconds), meaning 60KB/sec, or 1KB/frame, or ~128B/call? Obviously 0B/call is _better_ and I'm all for improvements where they make sense.
The thing that surprises me is that 60KB/sec is significant enough to cause noticeable improvements in the frequency of GCs.

^if I'm wrong about any of that please let me know

This is sounding like a performance tweak that's more specific to your application, at least the constraint of how many arguments to support. I think there's a better fix for your code that should be better both for CPU performance and memory overhead, and that is to not use embind on any calls you consider "hot". Embind makes the interface between JS and C++ much more natural, but it comes at a runtime cost. And in the case where we're tuning performance of hot calls, the lowest hanging fruit is going to be using a more bare-metal interface. [The most minimal abstraction we have is no abstraction at all](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#call-compiled-c-c-code-directly-from-javascript), so we can export raw wasm functions with zero code in between your C++ and JS.
So for something like:
```cpp
extern "C" { // to avoid C++ name mangling
  EMSCRIPTEN_KEEPALIVE // to export the function
  int foo(int a, int b) { ... }
}
```
we can call it like
```js
var x = Module["_foo"](1, 2);
```

Dyncalls are also not-fast, though not to the same degree as embind typically, and have less of a (conceptually) straightforward workaround. We should still try and remove as many `arguments` allocations in the tools as we can, I'm just trying to find the right balance here.

**Comment by google-feinberg** [comment]
OK. If I kill the embind.js changes, can we at least get the emscripten.py patch through?

**Comment by jgravelle-google** [comment]
I'm fine with that.
I have a slight preference for the arg-counting version. I can imagine a world where we extend that to include arg count metadata from binaryen for every function, and avoid using `arguments` for every exported wasm function, making it less dynCall-specific and more performant across the board.

**Comment by jgravelle-google** [comment]
> I have a slight preference for the arg-counting version.
@juj wdyt? It's speculative work so it might not materialize and then we would have this dyncall-specific-seeming code around for an indeterminate amount of time. Though I'm not opposed to that either.

**Comment by kripken** [comment]
The freetype and poppler failures here are due to an llvm update - merging in newest incoming to here will include the fix for that.

The browser test failures look real, though. One is that we increase code size, so the expected value `5651` in the test needs to be updated. The other is the runtime-misuse test, which checks that we give good errors on calling functions before the runtime is ready, stuff like that - I think this PR might actually be breaking the mechanism used for that. If so, an option is to not do this optimization if ASSERTIONS is on.

**Comment by jgravelle-google** [comment]
Thinking about this more I like the idea of reverting the emscripten.py change from https://github.com/emscripten-core/emscripten/pull/9922/commits/01af010663b4f7351a1124a22b4c7e32bb70ca21 and then emitting export arity metadata from binaryen, to avoid using `arguments` and generating junk for any wasm exports. Do we have anything in the testsuite to benchmark memory pressure, or is that best tested by profiling larger programs?

**Comment by kripken** [comment]
I don't think we have any JS memory pressure tests currently.

Note that a downside to exporting the arity and then emitting the full list of arguments is that if there are many arguments then we'd be increasing code size. (However, that might only be a problem in dynamic linking, which we need to optimize more anyhow.) But that might be a good path, yeah.

**Comment by google-feinberg** [comment]
Hi, folks. All tests pass.

**Comment by Akaricchi** [comment]
It would be nice to also handle this for `Asyncify.instrumentWasmExports` somehow: https://github.com/emscripten-core/emscripten/blob/incoming/src/library_async.js#L619

**Comment by google-feinberg** [comment]
> It would be nice to also handle this for `Asyncify.instrumentWasmExports` somehow: https://github.com/emscripten-core/emscripten/blob/incoming/src/library_async.js#L619

Can that come in some other pull request? As far as I know, our project doesn't use that, and it would be tough for me to understand enough to change.

**Comment by google-feinberg** [comment]
Any chance of a merge on this? It really makes a big difference in garbage production.

File: tests/core/test_longjmp_i64.c [hunk 3]
```diff
@@ -14,7 +16,7 @@
 jmp_buf jb;
 
 __attribute__((noinline)) void foo(int64_t x) {
-  printf("foo: %lld.\n", x);
+  printf("foo: %" PRId64 ".\n", x);
```
Comment (jgravelle-google) [inline tests/core/test_longjmp_i64.c]: This feels unrelated, weird that you had to do this. It seems more-correct but I'm wondering what happened there.

Comment (google-feinberg) [reply tests/core/test_longjmp_i64.c]: That test won't compile for me without making that change.
